File: Ex1211_AVLTree/AVLTree.cpp
#include <iostream>

using namespace std;

typedef int ElementType;

typedef struct AvlNode
{
	ElementType Element;
	AvlNode *Left;
	AvlNode *Right;
	int Height;
}*Position, *AvlTree;

AvlTree MakeEmpty(AvlTree T);
Position Find(ElementType x, const AvlTree T);
Position FindMin(const AvlTree T);
Position FindMax(const AvlTree T);
ElementType Retrieve(const Position P);
AvlTree Insert(ElementType x, AvlTree T);
Position SingleRotateWithLeft(Position K2);
Position SingleRotateWithRight(Position K2);
Position DoubleRotateWithLeft(Position K3);
Position DoubleRotateWithRight(Position K3);


int main(void)
{
	AvlTree T = nullptr;
	T = Insert(1, T);
	T = Insert(2, T);
	T = Insert(3, T);
	T = Insert(4, T);
	T = Insert(5, T);
	T = Insert(6, T);
	T = Insert(7, T);

	Position P = Find(5, T);
	int x = Retrieve(P);
	cout << "Position: " << P << endl;
	cout << "Element: " << x << endl;
	
	MakeEmpty(T);
	return 0;
}

AvlTree MakeEmpty(AvlTree T)
{
	if (T != nullptr)
	{
		MakeEmpty(T->Left);
		MakeEmpty(T->Right);

		delete T;
	}
	return nullptr;
}

Position Find(ElementType x, const AvlTree T)
{
	Position ptr = T;
	while (ptr != nullptr)
	{
		if (x < ptr->Element)
		{
			ptr = ptr->Left;
		}
		else if (x > ptr->Element)
		{
			ptr = ptr->Right;
		}
		else return ptr;
	}
	return nullptr;
}

Position FindMin(const AvlTree T)
{
	Position ptr = T;
	if (ptr != nullptr)
	{
		while (ptr->Left != nullptr)
		{
			ptr = ptr->Left;
		}
	}
	return ptr;
}

Position FindMax(const AvlTree T)
{
	Position ptr = T;
	if (ptr != nullptr)
	{
		while (ptr->Right != nullptr)
		{
			ptr = ptr->Right;
		}
	}
	return ptr;
}

ElementType Retrieve(const Position P)
{
	if (P == nullptr)
	{
		throw "Empty Position";
	}
	else
	{
		return P->Element;
	}
}

int Height(const Position P)
{
	if (P == nullptr)
	{
		return -1;
	}
	else return P->Height;
}

AvlTree Insert(ElementType x, AvlTree T)
{
	if (T == nullptr)
	{
		T = new AvlNode;
		T->Element = x;
		T->Height = 0;
		T->Left = T->Right = nullptr;
	}
	else if (x < T->Element)
	{
		T->Left = Insert(x, T->Left);
		if (Height(T->Left) - Height(T->Right) == 2)
		{
			if (x < T->Left->Element)
			{
				T = SingleRotateWithLeft(T);
			}
			else
			{
				T = DoubleRotateWithLeft(T);
			}
		}
	}
	else if (x > T->Element)
	{
		T->Right = Insert(x, T->Right);
		if (Height(T->Right) - Height(T->Left) == 2)
		{
			if (x > T->Right->Element)
			{
				T = SingleRotateWithRight(T);
			}
			else
			{
				T = DoubleRotateWithRight(T);
			}
		}
	}

	T->Height = Height(T->Left) > Height(T->Right) 
		? Height(T->Left) + 1 : Height(T->Right) + 1;

	return T;
}

Position SingleRotateWithLeft(Position K2)
{
	Position K1 = K2->Left;
	K2->Left = K1->Right;
	K1->Right = K2;

	K2->Height = Height(K2->Left) > Height(K2->Right)
		? Height(K2->Left) + 1 : Height(K2->Right) + 1;
	K1->Height = Height(K1->Left) > Height(K2)
		? Height(K1->Left) + 1 : Height(K2) + 1;

	return K1;
}

Position SingleRotateWithRight(Position K2)
{
	Position K1 = K2->Right;
	K2->Right = K1->Left;
	K1->Left = K2;

	K2->Height = Height(K2->Left) > Height(K2->Right)
		? Height(K2->Left) + 1 : Height(K2->Right) + 1;
	K1->Height = Height(K1->Left) > Height(K2)
		? Height(K1->Left) + 1 : Height(K2) + 1;

	return K1;
}

Position DoubleRotateWithLeft(Position K3)
{
	K3->Left = SingleRotateWithRight(K3->Left);

	return SingleRotateWithLeft(K3);
}

Position DoubleRotateWithRight(Position K3)
{
	K3->Right = SingleRotateWithLeft(K3->Right);

	return SingleRotateWithRight(K3);
}

File: Ex1211_AVLTree/AVLTree.java
class TestAVLTree {
    public static void main(String[] args) {
        AVLTree<Integer> tree = new AVLTree<>();

        tree.insert(new Integer[]{1, 3, 5, 2, 6, 7, 4});

        System.out.println("find(5): " + tree.find(5).getData());
        System.out.println("findMin(): " + tree.findMin().getData());
        System.out.println("findMax(): " + tree.findMax().getData());
    }
}

public class AVLTree<E extends Comparable<E>> {

    public static class AVLNode<E extends Comparable<E>> {
        private final E data;
        private AVLNode<E> leftChild;
        private AVLNode<E> rightChild;
        private int height;

        public AVLNode(E data) {
            this(data, null, null);
        }

        public AVLNode(E data, AVLNode<E> leftChild
                , AVLNode<E> rightChild) {
            this.data = data;
            this.leftChild = leftChild;
            this.rightChild = rightChild;
            this.height = 0;
        }

        public E getData() {
            return data;
        }

        public static int getHeight(AVLNode<?> node) {
            return node == null ? -1: node.height;
        }

        private void setHeight() {
            this.height = Math.max(getHeight(leftChild)
                    , getHeight(rightChild)) + 1;
        }

        // "this" is k2
        private AVLNode<E> singleRotateWithLeft() {
            AVLNode<E> k1 = this.leftChild;
            this.leftChild = k1.rightChild;
            k1.rightChild = this;

            this.setHeight();
            k1.setHeight();

            return k1;
        }

        // "this" is k2
        private AVLNode<E> singleRotateWithRight() {
            AVLNode<E> k1 = this.rightChild;
            this.rightChild = k1.leftChild;
            k1.leftChild = this;

            this.setHeight();
            k1.setHeight();
            return k1;
        }

        // "this" is k3
        private AVLNode<E> doubleRotateWithLeft() {
            this.leftChild = this.leftChild.singleRotateWithRight();

            return this.singleRotateWithLeft();
        }

        // "this" is k3
        private AVLNode<E> doubleRotateWithRight() {
            this.rightChild = this.rightChild.singleRotateWithLeft();

            return this.singleRotateWithRight();
        }

        private static final int ALLOW_BALANCE = 1;

        private AVLNode<E> balance(AVLNode<E> node) {
            if (node == null) {
                return null;
            }

            if (getHeight(node.leftChild) - getHeight(node.rightChild)
                    > ALLOW_BALANCE) {
                if (getHeight(node.leftChild.leftChild)
                        >= getHeight(node.leftChild.rightChild)) {
                    node = node.singleRotateWithLeft();
                }
                else {
                    node = node.doubleRotateWithLeft();
                }
            }
            else if (getHeight(node.rightChild) - getHeight(node.leftChild)
                    > ALLOW_BALANCE) {
                if (getHeight(node.rightChild.rightChild)
                        >= getHeight(node.rightChild.leftChild)) {
                    node = node.singleRotateWithRight();
                }
                else {
                    node = node.doubleRotateWithRight();
                }
            }

            node.setHeight();
            return node;
        }

        public AVLNode<E> insert(E data, AVLNode<E> node) {
            if (node == null) {
                return new AVLNode<>(data);
            }

            int compareResult = data.compareTo(node.data);

            if (compareResult < 0) {
                node.leftChild = insert(data, node.leftChild);
            }
            else if (compareResult > 0) {
                node.rightChild = insert(data, node.rightChild);
            }
            else {
                // Duplicate, do nothing
            }

            return balance(node);
        }

        public AVLNode<E> find(E data) {
            if (data == null) {
                throw new NullPointerException();
            }

            AVLNode<E> node = this;
            while (node != null) {
                if (data.compareTo(node.data) < 0) {
                    node = node.leftChild;
                }
                else if (data.compareTo(node.data) > 0) {
                    node = node.rightChild;
                }
                else return node;
            }
            return null;
        }

        public AVLNode<E> findMin() {
            AVLNode<E> node = this;
            while (node.leftChild != null) {
                node = node.leftChild;
            }
            return node;
        }

        public AVLNode<E> findMax() {
            AVLNode<E> node = this;
            while (node.rightChild != null) {
                node = node.rightChild;
            }
            return node;
        }
    }

    private AVLNode<E> root = null;

    public AVLTree() {}

    public void insert(E data) {
        if (root == null) {
            root = new AVLNode<>(data);
        }
        else {
            root = root.insert(data, root);
        }
    }

    public void insert(E[] array) {
        if (array == null) {
            throw new NullPointerException();
        }
        for (E data: array) {
            insert(data);
        }
    }

    public boolean isEmpty() {
        return root == null;
    }

    private void checkEmpty() {
        if (isEmpty()) {
            throw new NullPointerException("The tree is empty.");
        }
    }

    public AVLNode<E> find(E data) {
        checkEmpty();
        return root.find(data);
    }

    public AVLNode<E> findMin() {
        checkEmpty();
        return root.findMin();
    }

    public AVLNode<E> findMax() {
        checkEmpty();
        return root.findMax();
    }
}
